### PR TITLE
feat:Passcode Implementation

### PR DESCRIPTION
**Pull request by dilpreet2028**
**Note: There would be waiting time in gif in order to show the feature of asking passcode after 30 seconds (for demo purpose I set the threshold to 5 seconds)**

ss

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

**Comment by mayank-kgp** [comment]
@dilpreet96 Your UI of passcode activity looks very good. I will do the same.  :)

**Comment by dilpreet2028** [comment]
Thanks @mayank-kgp but there are some changes which rajan told me to do.
I will be using this library https://github.com/glomadrian/material-code-input for the input field and we will not be using the android keyboard so we need to make a gridview of all buttons 0-9. We can discuss this over a hangouts call if you want

**Comment by mayank-kgp** [comment]
@dilpreet96 Please come on slack.

**Comment by dilpreet2028** [comment]
@therajanmaurya changes made.I have not used the library which we finalised ,there were some problem with it so I made it from scratch.
@mayank-kgp have a look if you are going to implement it in android client.

**Comment by therajanmaurya** [comment]
@dilpreet96 Great work 👍. I will review today.

File: app/build.gradle [hunk 1]
```diff
@@ -70,6 +70,7 @@ dependencies {
     compile "com.android.support:cardview-v7:$rootProject.supportLibraryVersion"
     compile "com.google.android.gms:play-services-maps:$rootProject.playServicesVersion"
     compile "com.android.support:recyclerview-v7:$rootProject.supportLibraryVersion"
+    compile "com.android.support:gridlayout-v7:$rootProject.supportLibraryVersion"
```
Comment (therajanmaurya) [inline app/build.gradle]: why do you need `support:gridlayout`, I mean you can do same thing using `support:recyclerview`
` GridLayoutManager mLayoutManager = new GridLayoutManager(getActivity(), grid_count);`

**Review by therajanmaurya — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 1 above)
Comment (dilpreet2028) [reply app/build.gradle]: As the number buttons were static so I used Gridlayout for this task : https://github.com/dilpreet96/self-service-app/blob/bbc05af6a6814a4e568e59451e9b709981cae7f6/app/src/main/res/layout/activity_pass_code.xml#L55
We can use TableLayout also shall I update the PR with tablelayout?

Comment (therajanmaurya) [reply app/build.gradle]: yeah do that, if it is not affecting the UI and please add a Button to show and hide passcode

Comment (dilpreet2028) [reply app/build.gradle]: @therajanmaurya changes made, please review. I have updated the screenshots also.

**Comment by dilpreet2028** [comment]
@droidchef @therajanmaurya update the PR according to the suggestion
Added a Minimum threshold of 30 seconds after which user would be asked to enter passcode when he returns back and Added network check too.

**Comment by therajanmaurya** [comment]
@dilpreet96 Add screen shot too.

**Comment by dilpreet96** [comment]
@therajanmaurya I have updated the description part with new gif

File: app/src/main/java/org/mifos/selfserviceapp/ui/activities/HomeActivity.java [hunk 2]
```diff
@@ -76,7 +75,7 @@ public void onCreate(Bundle savedInstanceState) {
 
         setToolbarTitle(getString(R.string.home));
 
-        clientId = getIntent().getExtras().getLong(Constants.CLIENT_ID);
+        clientId = preferencesHelper.getClientId();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/selfserviceapp/ui/activities/HomeActivity.java:78]: @dilpreet96 Good work 👍, This is the correct way of doing this, During application period everyone was contributing so I did not get the chance to fix this. Please if you find this anywhere in the project do same.

File: app/src/main/java/org/mifos/selfserviceapp/ui/activities/LoginActivity.java [hunk 3]
```diff
@@ -79,11 +87,8 @@ public void hideProgress() {
     }
 
     @Override
-    public void showClient(long clientId) {
-        Intent accountsActivityIntent = new Intent(this, HomeActivity.class);
-        accountsActivityIntent.putExtra(Constants.CLIENT_ID, clientId);
-        startActivity(accountsActivityIntent);
-        finish();
+    public void showClient() {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/selfserviceapp/ui/activities/LoginActivity.java]: This method name should be `showPassCodeActivity()`

**Review by therajanmaurya — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/selfserviceapp/ui/activities/LoginActivity.java (hunk 3 above)
Comment (dilpreet2028) [reply app/src/main/java/org/mifos/selfserviceapp/ui/activities/LoginActivity.java]: @therajanmaurya Changes made

**Review by therajanmaurya — APPROVED** [review]

File: app/src/main/java/org/mifos/selfserviceapp/ui/activities/HomeActivity.java (hunk 2 above)
Comment (zhangch1991425) [reply app/src/main/java/org/mifos/selfserviceapp/ui/activities/HomeActivity.java]: test

**Comment by zhangch1991425** [comment]
test

**Pull request merged**